services-core/src/solution_submission/retry.rs

```diff
@@ -116,18 +120,97 @@ fn new_gas_price_estimate(
-        let time_remaining = args
-            .target_confirm_time
-            .saturating_duration_since(self.now.instant_now());
+    async fn gas_price(&self, target_confirm_time: Instant) -> Result<f64> {
+        let time_remaining = target_confirm_time.saturating_duration_since(self.now.instant_now());
         // TODO: Use a more accurate gas limit once the gas estimators take that into account.
         self.gas_price_estimating
             .estimate_with_limits(SOLUTION_SUBMISSION_GAS_LIMIT as f64, time_remaining)
             .await
     }
 
+    async fn submit_solution(&self, args: &Args, gas_price: f64) -> RetryResult {
+        RetryResult::Submitted(
+            self.contract
+                .submit_solution(
+                    args.batch_index,
+                    args.solution.clone(),
+                    args.claimed_objective_value,
+                    U256::from_f64_lossy(gas_price),
+                    args.nonce,
+                )
+                .await,
+        )
+    }
+
+    async fn cancel(&self, gas_price: f64, nonce: U256) -> RetryResult {
+        let gas_price = U256::from_f64_lossy(gas_price);
+        log::debug!("cancelling transaction with gas price {}", gas_price);
+        let result = self.contract.send_noop_transaction(gas_price, nonce).await;
+        RetryResult::Cancelled(result.map(|_| ()))
+    }
+
+    // Yields the current gas price immediately and then every refresh interval.
+    fn gas_price_stream(
+        &self,
+        target_confirm_time: Instant,
+    ) -> impl FusedStream<Item = Result<f64>> + '_ {
+        stream::unfold(true, move |first_call| async move {
+            if !first_call {
+                self.async_sleep.sleep(GAS_PRICE_REFRESH_INTERVAL).await;
+            }
+            return Some((self.gas_price(target_confirm_time).await, false));
+        })
+    }
+
     async fn retry_(&self, args: Args, cancel_after: impl Future) -> RetryResult {
-        todo!()
+        log::debug!("starting retry with gas price cap {}", args.gas_price_cap);
+
+        let gas_price_stream = self.gas_price_stream(args.target_confirm_time);
+        // make useable in `select!`
+        let cancel_after = cancel_after.fuse();
+        futures::pin_mut!(cancel_after);
+        futures::pin_mut!(gas_price_stream);
+
+        // This struct keeps track of all the solution and cancellation futures. If we get a
+        // "nonce already used error" we continue running the other futures. We need to handle this
+        // case because we do not know which transactions will complete or fail or in which order we
+        // observe completion.
+        let mut first_match = FirstMatch::new(|result: &RetryResult| !is_transaction_error(result));
+        let mut last_used_gas_price = 0.0;
+        loop {
+            futures::select! {
+                // Unwrap because the stream never ends.
+                gas_price = gas_price_stream.next() => match gas_price.unwrap() {
```

Comment:
ubernit: Instead of a comment, maybe use `expect`?
```suggestion
                gas_price = gas_price_stream.next() => match gas_price.expect("stream should never end") {
```